import sys, os
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
from matplotlib.patches import Rectangle
from scipy.stats import chi2_contingency
import statistics
import matplotlib.patches as mpatches

import argparse

parser = argparse.ArgumentParser(description="This script generates a figure detailing the location of Core and Acessory genes in the chromosome.")
parser.add_argument("-o", "--orthofinder_dir", required=True, help="Directory containing the output of an OrthoFinder run")
parser.add_argument("-n", "--name", default="", help="A string for the title of the plot")
parser.add_argument("-c", '--csv', required=True, help="csv file generated by csv_landscape.py, with detailed per gene data for a single species")
parser.add_argument("-e", "--entry_list", default="", help="File containing the list of entries to consider for computing the pangenome")
parser.add_argument("-w", "--window_size", default=100000, help="Genome window size for various calculations")

args = parser.parse_args()

step = args.window_size
df = pd.read_csv(args.csv)
df = df.query("core_acc != 'other'")

print(df)

thefile = open(args.csv)
indexlist = df.index.values.tolist()

orthofinder_dir = args.orthofinder_dir
if orthofinder_dir[-1] == "/":
	orthofinder_dir = orthofinder_dir[:-1]
orthofile = orthofinder_dir + "/Orthogroups/Orthogroups.GeneCount.tsv"
orthonames = orthofinder_dir + "/Orthogroups/Orthogroups.tsv"
scoog_dir = orthofinder_dir + "/Single_Copy_Orthologue_Sequences/"
sppfile, spplist = args.entry_list , []
figname = args.name
if len(figname) > 0: 
	figname = ": " + figname
spppos, scolist = {}, []

def bluepicker(number, end):
	value = end - number
	midpoint = int(end/2)
	if value < midpoint:
		g = 1 - value/midpoint
		r = 1
		b = 1 - value/midpoint
	else:
		g = 0
		r = 1 - (value - midpoint)/midpoint
		b = 0
	return (r, g, b)
	
for i in os.listdir(scoog_dir):
	scolist.append(i.split(".")[0])

for i in open(sppfile):
	spplist.append(i[:-1])

pav_dict = {}
gene2og = {}

for line in open(orthofile, "r"):
	chunk = line[:-1].split("\t")
	if chunk[0] == "Orthogroup":
		for i in range(len(chunk)):
			if chunk[i] in spplist:
				spppos[i] = chunk[i]
	else:
		c = 0
		for i in range(len(chunk)):
			if i in spppos and int(chunk[i]) > 0:
				c = c + 1
		pav_dict[chunk[0]] = c

for line in open(orthonames, "r"):
	chunk = line[:-1].split("\t")
	if chunk[0] == "Orthogroup": continue
	else:
		for i in range(len(chunk)):
			if i in spppos and len(chunk[i]) > 1:
				for gene in chunk[i].replace(" ", "").split(","):
					gene2og[gene] = chunk[0]

fig = plt.figure(figsize=(150, 100))

chromodict = {}
for index, row in df.iterrows():
	chromodict[row["scaffold"]] = row["scaflen"]

scaf_names = []
scaf_fullnames = []
for k in chromodict.keys():
	scaf_names.append(k.split("_")[1])
	scaf_fullnames.append(k)

ax = fig.add_subplot(111)

ax.bar(scaf_names, chromodict.values(), color="ivory")
plt.ylim((0,7200000))
plt.xlabel('Scaffold', fontsize=18)
plt.ylabel('Scaffold length', fontsize=18)
plt.title('Distribution of core and accessory genes'+figname, fontsize=28)

for i in indexlist:
	n = df.loc[i, "gene"]
	if len(spplist) == pav_dict[gene2og[n]]:
		c1 = "ivory"
		if gene2og[n] in scolist:
			c2 = "navy"
		else:
			c2 = "deepskyblue"
	else:
		factor = pav_dict[gene2og[n]]
		c1 = bluepicker(factor, len(spplist))
		c2 = "ivory"
	ax.add_patch(Rectangle((scaf_fullnames.index(df.loc[i,"scaffold"])-0.2, df.loc[i,"start"]), 0.3, df.loc[i,"len"], color=c2, alpha=0.8))
	ax.add_patch(Rectangle((scaf_fullnames.index(df.loc[i,"scaffold"])+0.1, df.loc[i,"start"]), 0.3, df.loc[i,"len"], color=c1, alpha=0.8))

df['midpoint'] = (df['start'] + df['end']) / 2
total_core_genes = len(df[df['core_acc'] == 'core'])
total_acc_genes = len(df[df['core_acc'] == 'acc'])
total_genes = len(df)

for c in scaf_fullnames:
	region_start, region_end = 0, step
	dfc = df[df['scaffold'] == c]
	while region_start < chromodict[c]:
		genes_in_region = dfc[(dfc['midpoint'] >= region_start) & (dfc['midpoint'] < region_end)]
		core_count_in_region = len(genes_in_region[genes_in_region['core_acc'] == 'core'])
		acc_count_in_region = len(genes_in_region[genes_in_region['core_acc'] == 'acc'])
		total_genes_in_region = len(genes_in_region)
		expected_core_in_region = (total_core_genes / total_genes) * total_genes_in_region
		expected_acc_in_region = (total_acc_genes / total_genes) * total_genes_in_region
		observed = np.array([core_count_in_region, acc_count_in_region])
		expected = np.array([expected_core_in_region, expected_acc_in_region])
		chi2, p_value = chi2_contingency([observed, expected])[:2]
		gsum, sco_check = [], False
		for g in genes_in_region["gene"]:
			if g == "gene": continue
			if gene2og[g] in scolist: sco_check = True
			if pav_dict[gene2og[g]] == len(spplist): continue
			else:
				gsum.append(pav_dict[gene2og[g]])
		if len(gsum) == 0:
			gsum.append(0)
		average = statistics.fmean(gsum)
		
		#print(f'Observed core genes in region: {core_count_in_region}')
		#print(f'Observed accessory genes in region: {acc_count_in_region}')
		#print(f'Expected core genes in region: {expected_core_in_region}')
		#print(f'Expected accessory genes in region: {expected_acc_in_region}')
		#print(f'Chi-squared Statistic: {chi2}, P-value: {p_value}')
		
		farge = "grey"
		if p_value < 0.05 and core_count_in_region > acc_count_in_region:
			farge = "deepskyblue"
		if p_value < 0.05 and core_count_in_region < acc_count_in_region:
			farge = "tomato"
		ax.add_patch(Rectangle((scaf_fullnames.index(c)-0.4, region_start), 0.2, step, color=farge, alpha=0.8))
		#ax.text(scaf_fullnames.index(c)-0.35, region_start+(0.3*step), str(round(average,1)), fontsize=6)
		if sco_check == False:
			ax.add_patch(Rectangle((scaf_fullnames.index(c)-0.45, region_start), 0.04, step, color="deeppink", alpha=0.8))
		region_start, region_end = region_start+step , region_end+step
ax.add_patch(Rectangle((scaf_fullnames.index("TA10106M1_Scaffold02")-0.45, 3628802), 0.25, (3628802-3646776), color="lime", alpha=0.8))
ax.add_patch(Rectangle((scaf_fullnames.index("TA10106M1_Scaffold09")-0.45, 2173911), 0.25, (2173911-2204167), color="yellow", alpha=0.8))

import sys, os
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
from matplotlib.patches import Rectangle
from scipy.stats import chi2_contingency
import statistics
import matplotlib.patches as mpatches


def bluepicker(number, end):
	value = end - number
	midpoint = int(end/2)
	if value < midpoint:
		g = 1 - value/midpoint
		r = 1
		b = 1 - value/midpoint
	else:
		g = 0
		r = 1 - (value - midpoint)/midpoint
		b = 0
	return (r, g, b)


start, end = 1, 17

for i in range(start, end):
	bignum = 100000
	f = (i*bignum)+5000000
	c = bluepicker(i, end)
	ax.add_patch(Rectangle((10,f), 0.4, bignum, color=c))
	ax.text(9.8,f, str(i), fontsize=10)
ax.add_patch(Rectangle((9,f), 0.4, bignum, color="deepskyblue"))
ax.text(8.7,f, str("Core"), fontsize=10)
ax.add_patch(Rectangle((9,f-bignum), 0.4, bignum, color="navy"))
ax.text(8.7,f-bignum, str("SCOs"), fontsize=10)

# Showing the plot
plt.show()

